google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,702 @@
+namespace {
+
+using ::google::cloud::testing_util::IsOk;
+using ::testing::_;
+using ::testing::StrictMock;
+using ::testing::MockFunction;
+using ::testing::Return;
+using ::testing::ByMove;
+
+using ::google::cloud::pubsublite_internal::MockRetryPolicy;
+using ::google::cloud::pubsublite_internal::MockBackoffPolicy;
+
+using ms = std::chrono::milliseconds;
+const unsigned int kFutureWaitMs = 100;
+
+struct FakeRequest {
+  std::string key;
+};
+
+struct FakeResponse {
+  std::string key;
+  std::string value;
+};
+
+// Used as part of the EXPECT_CALL() below:
+bool operator==(FakeRequest const& lhs, FakeRequest const& rhs) {
+  return lhs.key == rhs.key;
+}
+
+bool operator==(FakeResponse const& lhs, FakeResponse const& rhs) {
+  return lhs.key == rhs.key && lhs.value == rhs.value;
+}
+
+using AsyncReaderWriter = ::google::cloud::pubsublite_internal::MockAsyncReaderWriter<
+    FakeRequest,
+    FakeResponse>;
+
+using AsyncReadWriteStreamReturnType =
+std::unique_ptr<AsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+using ResumableAsyncReadWriteStream = std::unique_ptr<
+    ResumableAsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+const Status kFailStatus = Status(StatusCode::kUnavailable, "Unavailable");
+const FakeResponse kBasicResponse = FakeResponse{"key0", "value0_0"};
+const FakeResponse kBasicResponse1 = FakeResponse{"key0", "value0_1"};
+const FakeRequest kBasicRequest = FakeRequest{"key0"};
+
+class ResumableAsyncReadWriteStreamingRpcTest : public ::testing::Test {
+ protected:
+  ResumableAsyncReadWriteStreamingRpcTest() : backoff_policy_{
+      std::make_shared<StrictMock<MockBackoffPolicy>>()}, stream_{
+      MakeResumableAsyncStreamingReadWriteRpcImpl<FakeRequest, FakeResponse>(
+          [this]() { return retry_policy_factory_.Call(); },
+          backoff_policy_,
+          [this](std::chrono::duration<double> t) { return sleeper_.Call(t); },
+          [this]() { return stream_factory_.Call(); },
+          [this](AsyncReadWriteStreamReturnType stream) {
+            return initializer_.Call(std::move(stream));
+          })} {
+  }
+
+  StrictMock<MockFunction<future<void>(std::chrono::duration<double>)>>
+      sleeper_;
+  StrictMock<MockFunction<AsyncReadWriteStreamReturnType()>> stream_factory_;
+  StrictMock<MockFunction<std::unique_ptr<MockRetryPolicy>()>>
+      retry_policy_factory_;
+  StrictMock<MockFunction<future<StatusOr<AsyncReadWriteStreamReturnType>>(
+      AsyncReadWriteStreamReturnType)>>
+      initializer_;
+  std::shared_ptr<StrictMock<MockBackoffPolicy>> backoff_policy_;
+  ResumableAsyncReadWriteStream stream_;
+
+  void InitializerBehavior(unsigned int times) {
+    EXPECT_CALL(initializer_, Call).Times(times).WillRepeatedly([](
+        AsyncReadWriteStreamReturnType stream) {
+      return make_ready_future(
+          StatusOr<AsyncReadWriteStreamReturnType>(std::move(stream)));
+    });
+  }
+
+  void NoRetryInvocation(unsigned int times) {
```

Comment:
Inline this everywhere and make Return(ByMove(absl::make_unique<StrictMock<MockRetryPolicy>>()))